doc/developer/design/20250415_large_select_result_size.md

```diff
@@ -0,0 +1,174 @@
+
+We think it is easier to lift the result size limitation for streamable
+queries, so want to approach that first, but there are ways we can lift the
+limitation for non-streamable queries as well, for example by moving the post
+processing to the cluster side or by implementing an approach where we can sort
+and merge large results within bounded memory on the `environmentd` side.
+
+## Technical Context
+
+There are two ways of getting query results and sending them from the cluster
+back to `environmentd`:
+
+1. Extracting from an arrangement (the data structure backing indexes)
+2. Extracting directly from persist
+
+This leads to there being at least three ways a SELECT can be executed from the
+`environmentd` side:
+
+1. Fast-path SELECT: there is an existing arrangement (index) that we can
+   extract our result from.
+2. Slow-path SELECT: in order to get our result we first need to render a
+   temporary dataflow with an index that will contain the results. Then we can
+   use the same logic as a fast-path SELECT to extract the result. Afterwards
+   the temporary dataflow is torn down again.
+3. Fast-path persist SELECT: The query is simple enough that we can directly
+   extract it from a persist shard. With potentially some amount of
+   map-filter-project applied. This happens on the `clusterd` side and is
+   shipped back using the same mechanism that is used for shipping back other
+   SELECT results.
+
+1 and 2 here imply that the result must fit into memory on the cluster side,
+unless we want to, say, change how results can be extracted on the cluster
+side.
+
+## Goals
+
+The user facing goal is:
+
+- Lift result-size limitation for streamable queries
+
+And we have technical goals as well, enabling potential future work and making
+sure we have the right abstractions in place:
+
+- Create re-usable building blocks for shipping data/results out-of-band, these
+  can then be used in other places, such as for SUBSCRIBE
+- Make persist Batches more of a "first-class" citizen, eventually allowing us
+  to decouple where and how batches are written from where and how they are
+  appended to a shard
+
+## Non-Goals
+
+- Lift limitation on result having to fit into cluster memory: for non-persist
+  queries, the result is still first staged in an arrangement before we read it
+  out and send it to `environmentd`. Or we read out of an existing arrangement
+- Change how results are extracted from "the dataflows" on the cluster side
+- Lift result-size limitation for non-streamable queries
+- Change result-size limitations for SUBSCRIBE
+- Change how we ship around and append data to shards, when handling INSERT
+- Change how we ship around and append data for sources
+
+## Implementation
+
+Purposefully high level, but the rough idea is that we a) need to stop sending
+large amounts of data through the compute protocol, and b) need to stop
+materializing large results in `environmentd` memory.
```

Comment:
How much of a problem is it to send large amounts of data through the compute protocol? an alternative design could be implemented entirely within envd. It could receive the data from replicas, and instead of materializing it in memory, it could write to files or blob storage, and stream the results to the client from there. Just pointing out the option!

Reply:
I would say it is a problem, and not just for peeks but also for subscribes, which you found in https://github.com/MaterializeInc/database-issues/issues/9264.

An alternative would be to add a protocol between `clusterd` and `envd` for sending back these large responses, which doesn't block the "main" protocol. But I'd say the design proposed here is already kind of that, but it uses persist blob storage as the intermediary, and I think that's easier to implement compared to introducing a new protocol.